### ISSUE TITLE
Span (or generalized) foreach support

### ISSUE THREAD

**Issue by stephentoub** [dotnet/csharplang#1085]
`T[]` exposes a `GetEnumerator` method so that it can be `foreach`'d, but it's also specially recognized by the C# compiler, with IL emitted that's equivalent to a `for` loop from `0` to `array.Length`.

`Span<T>` is meant to be usable in situations where `T[]` is usable, but `foreach` doesn't currently work with `Span<T>` (or `ReadOnlySpan<T>`).  We can add a `GetEnumerator` to `Span<T>` that returns a `ref struct Enumerator`, but a) it would have to be an instance method as `foreach` doesn't currently support extension methods, and we're trying to avoid such instance methods on these core generic types in order to avoid bloat in the generated asm (especially for AOT), and b) it wouldn't have the same optimized IL that arrays benefit from.

__Proposal:__
Support `Span<T>` and `ReadOnlySpan<T>` in `foreach`, such that:
```C#
foreach (T item in span)
{
    ...
}
```
is equivalent to:
```C#
for (int i = 0; i < span.Length; i++)
{
    T item = span[i];
    ...
}
```

The compiler could either specially recognize `Span<T>` and `ReadOnlySpan<T>`, or alternatively we could have an attribute the compiler recognizes, like:
```C#
[ForeachWithLengthAndIndexer]
```
Such an attribute could be applied to a type like `Span<T>`, and when the compiler saw such an attribute, it would generate for a `foreach` loop code that uses the target's `Length` and `get_Item[int]` indexer.

cc: @MadsTorgersen, @VSadov

**Comment by CyrusNajmabadi** [comment]
If we were going to do work in the compiler to make this situation, would it suffice to have a more general solution that addresses things by tackling this issue:

> it would have to be an instance method as foreach doesn't currently support extension methods,

I.e. if foreach could pick up extension methods, would that be a suitable solution?  Or would that not be good enough (for example, the perf wouldn't be as good as a direct .Length + .get_Item solution).

--

It would be great to just relax foreach and make it more pattern-y, like we've done with async/await.  It would make sense anyways given that foreach is *already* patterny, just with an 'instance member' restriction.

**Comment by VSadov** [comment]
Here is some info on how it currently works: (just for perspective) 

`foreach` is pattern based and requires `GetEnumerator` of a particular shape with some minor variations. Form the pattern compiler can infer the type of the iterator and the type of the iteration element. (it is not hardcoded that iteration element for `string` is `char`).
The special casing of `T[]` and `string` are optional optimizations. They are not required by the spec - as long as your compiler understands foreach pattern it is a correct C# compiler. In fact some runtimes may not have enough support for optimized `foreach` and that is not a failure. If `string` does not have `Chars` property (yes, such runtimes exist), then compiler emits unoptimized loop. What else could it do?

Ideally, `Span<T>` would work the same way - it would implement foreach pattern and compiler will be free to lower `foreach` loops into `for` loops as a matter of optimization, but would not required to do so.

**Comment by stephentoub** [comment]
> Ideally, Span<T> would work the same way - it would implement foreach pattern and compiler will be free to lower foreach loops into for loops as a matter of optimization, but would not required to do so.

If that's where we land, can we enable the pattern to be satisfied with an extension GetEnumerator?

**Comment by CyrusNajmabadi** [comment]
> If that's where we land, can we enable the pattern to be satisfied with an extension GetEnumerator?

This seems like this would be a very low cost way to solve the problem, and would make the language more general and consistent with how it handles pattern based approaches to things.

**Comment by VSadov** [comment]
There could be valid reasons why GetEnumerator cannot be an extension or the limitation is just to simplify lookup - I am not sure. I do no think it would be a breaking change. Existing instance GetEnumerator would win and existing cases would continue working. 

I wonder why extension is better than a member? 
Do you see a scenario where you do not want GetEnumerator be in scope or want an alternative one (with a different element typo or something). 
Basically why making it an extension point vs baking the capability into the type?

I am also not sure about JIT/AOT concerns - if code is not called (due to optimized binary just using indexing) it would not be jitted, right?

**Comment by stephentoub** [comment]
> I am also not sure about JIT/AOT concerns

@jkotas, have I misunderstood the concerns about making things instance members?

**Comment by jkotas** [comment]
> I am also not sure about JIT/AOT concerns - if code is not called (due to optimized binary just using indexing) it would not be jitted, right?

Our AOT technologies pretty much always pre-compile all members of the generic type.

**Comment by jkotas** [comment]
Span / Memory is likely to be used with much fewer different instantiations than say ValueTuple or Dictionary. I think we can live what the code bloat for Span / Memory.

**Comment by VSadov** [comment]
So it could be:
1)  GetEnumerator directly on Span. 
Could be pretty efficient. The enumerator can be just two references to the current and the last element. 
Compiler can optimize that into indexing if that is better.
No language changes necessary.

2) GetEnumerator as extension method. 
Could still be efficiently implemented. I am not sure compiler can easily optimize this though. Since it is an extension we are less certain what extension does.
Need to tweak lookup rules for `GetEnumerator` which has much broader impact on the language than needed here.

3) Some other pattern that tells "use `Length` and indexer". 
Interesting and could be useful to things like `List<T>` as well. We generally do not optimize those. In particular List<T> iterator behaves differently from indexing if list mutates between iterations. 
But if type says "go ahead iterate via indexing", we might be ok. Except for silent change in behavior when changing between old/new compilers.

I think so far the `#1` is my preferred choice - implement `GetEnumerator` as a member and also optimize in C# compiler to use indexing, if that gives any advantages.

**Comment by CyrusNajmabadi** [comment]
> I am not sure compiler can easily optimize this though. Since it is an extension we are less certain what extension does.

Would we really need compiler optimization here?  I would think the jit would be able to optimize the heck out of this.  At the very least, i'd want to get some numbers showing that the compiler approach really did squeeze out extra perf here over the runtime using a specialized method over ```Span<T>```.

> Need to tweak lookup rules for GetEnumerator which has much broader impact on the language than needed here.

That's a positive in my book (vs '3' at least).  I'd much rather us make an existing system better, and absorb this and other cases, versus needing to specialize the language/compiler here for just ```Span<T>```.

I do agree though that if this can just be done with '1' then this is effectively a non-issue.

**Comment by jnm2** [comment]
@VSadov 

> I wonder why extension is better than a member?
> Do you see a scenario where you do not want GetEnumerator be in scope or want an alternative one (with a different element type or something).
> Basically why making it an extension point vs baking the capability into the type?

The thing that I keep waiting for is https://github.com/dotnet/roslyn/issues/16159.

**Comment by stephentoub** [comment]
> i'd want to get some numbers showing that the compiler approach really did squeeze out extra perf here

There's a reason the C# compiler generates the code it does for arrays... it makes a significant difference.

Here's a ref struct Enumerator for span:
https://github.com/dotnet/coreclr/pull/14922
With that we can foreach.

Here's a little microbenchmark:
```C#
using System;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;

class Program
{
    static void Main()
    {
        Span<int> span = Enumerable.Range(0, 1_000_000).ToArray();
        var sw = new Stopwatch();
        while (true)
        {
            const int Iters = 1_000;

            sw.Restart();
            for (int i = 0; i < Iters; i++) Sum_For(span);
            sw.Stop();
            Console.WriteLine(sw.Elapsed.TotalMilliseconds);

            sw.Restart();
            for (int i = 0; i < Iters; i++) Sum_ForEach(span);
            sw.Stop();
            Console.WriteLine(sw.Elapsed.TotalMilliseconds);

            Console.WriteLine();
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Sum_For(Span<int> span)
    {
        int sum = 0;
        for (int i = 0; i < span.Length; i++) sum += span[i];
        return sum;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Sum_ForEach(Span<int> span)
    {
        int sum = 0;
        foreach (int i in span) sum += i;
        return sum;
    }
}
```
Example output on my machine:
```
c:\Users\stoub\Desktop\CoreClrTest>corerun test.exe
536.8486
1779.5118

517.6757
1758.5095

620.2551
1894.9973

597.6054
1824.4725

517.2217
1744.8355
```

Here's the asm for the Sum_For method, which has the IL we'd want the C# compiler to generate.
```
        int sum = 0;
00007FF9A4CA1FB1  mov         eax,dword ptr [rcx]  
00007FF9A4CA1FB3  mov         edx,dword ptr [rcx+8]  
00007FF9A4CA1FB6  xor         ecx,ecx  
        for (int i = 0; i < span.Length; i++) sum += span[i];
00007FF9A4CA1FB8  xor         r8d,r8d  
00007FF9A4CA1FBB  test        edx,edx  
00007FF9A4CA1FBD  jle         00007FF9A4CA1FCE  
00007FF9A4CA1FBF  movsxd      r9,r8d  
00007FF9A4CA1FC2  add         ecx,dword ptr [rax+r9*4]  
00007FF9A4CA1FC6  inc         r8d  
00007FF9A4CA1FC9  cmp         r8d,edx  
00007FF9A4CA1FCC  jl          00007FF9A4CA1FBF  
        return sum;
00007FF9A4CA1FCE  mov         eax,ecx  
00007FF9A4CA1FD0  ret  
```
and here's the asm for the Sum_ForEach method:
```
        int sum = 0;
00007FF9A4CA5962  sub         rsp,58h  
00007FF9A4CA5966  vzeroupper  
00007FF9A4CA5969  mov         rsi,rcx  
00007FF9A4CA596C  lea         rdi,[rsp+28h]  
00007FF9A4CA5971  mov         ecx,0Ch  
00007FF9A4CA5976  xor         eax,eax  
00007FF9A4CA5978  rep stos    dword ptr [rdi]  
00007FF9A4CA597A  mov         rcx,rsi  
00007FF9A4CA597D  xor         eax,eax  
        foreach (int i in span) sum += i;
00007FF9A4CA597F  mov         rdx,qword ptr [rcx]  
00007FF9A4CA5982  mov         ecx,dword ptr [rcx+8]  
00007FF9A4CA5985  xor         r8d,r8d  
00007FF9A4CA5988  lea         r9,[rsp+28h]  
00007FF9A4CA598D  vxorpd      xmm0,xmm0,xmm0  
00007FF9A4CA5992  vmovdqu     xmmword ptr [r9],xmm0  
00007FF9A4CA5997  mov         qword ptr [r9+10h],r8  
00007FF9A4CA599B  lea         r8,[rsp+28h]  
00007FF9A4CA59A0  mov         qword ptr [r8],rdx  
00007FF9A4CA59A3  mov         dword ptr [r8+8],ecx  
00007FF9A4CA59A7  xor         edx,edx  
00007FF9A4CA59A9  mov         dword ptr [rsp+38h],edx  
00007FF9A4CA59AD  mov         dword ptr [rsp+3Ch],edx  
00007FF9A4CA59B1  vmovdqu     xmm0,xmmword ptr [rsp+28h]  
00007FF9A4CA59B8  vmovdqu     xmmword ptr [rsp+40h],xmm0  
00007FF9A4CA59BF  mov         rdx,qword ptr [rsp+38h]  
00007FF9A4CA59C4  mov         qword ptr [rsp+50h],rdx  
00007FF9A4CA59C9  jmp         00007FF9A4CA59D1  
00007FF9A4CA59CB  mov         edx,dword ptr [rsp+54h]  
00007FF9A4CA59CF  add         eax,edx  
00007FF9A4CA59D1  mov         edx,dword ptr [rsp+50h]  
00007FF9A4CA59D5  cmp         edx,dword ptr [rsp+48h]  
00007FF9A4CA59D9  jge         00007FF9A4CA5A07  
00007FF9A4CA59DB  mov         edx,dword ptr [rsp+50h]  
00007FF9A4CA59DF  lea         rcx,[rsp+40h]  
00007FF9A4CA59E4  lea         r8,[rsp+40h]  
00007FF9A4CA59E9  lea         r9d,[rdx+1]  
00007FF9A4CA59ED  mov         dword ptr [rsp+50h],r9d  
00007FF9A4CA59F2  cmp         edx,dword ptr [r8+8]  
00007FF9A4CA59F6  jae         00007FF9A4CA5A0E  
00007FF9A4CA59F8  mov         r8,qword ptr [r8]  
00007FF9A4CA59FB  movsxd      rdx,edx  
00007FF9A4CA59FE  mov         edx,dword ptr [r8+rdx*4]  
00007FF9A4CA5A02  mov         dword ptr [rcx+14h],edx  
00007FF9A4CA5A05  jmp         00007FF9A4CA59CB  
00007FF9A4CA5A07  add         rsp,58h  
00007FF9A4CA5A0B  pop         rsi  
00007FF9A4CA5A0C  pop         rdi  
00007FF9A4CA5A0D  ret  
00007FF9A4CA5A0E  call        00007FFA048EF2C0  
00007FF9A4CA5A13  int         3  
```
Could the JIT do a better job here?  Probably (cc: @AndyAyersMS).  But the C# compiler could also make its job a heck of a lot easier by emitting the IL we'd like from the start.

**Comment by alrz** [comment]
Relates to https://github.com/dotnet/csharplang/issues/1046 (ref in foreach)

We could just require "ref indexer and length property" to make that work for other types. (non-array)

However, we might require https://github.com/dotnet/csharplang/issues/476 in turn.

**Comment by KrzysztofCwalina** [comment]
> I think so far the #1 is my preferred choice - implement GetEnumerator as a member and also optimize in C# compiler to use indexing, if that gives any advantages.

If we think we will have the compiler implementation, does it make sense to add GetEnumerator and Enumerator types? The types/code will be just dead code bloating assemblies, especially in AOT runtimes. We have been very careful to not bloat Span<T> (e.g. moved many methods to extenssions) so that the AOT images are small.

**Comment by VSadov** [comment]
There is a language spec and there are compiler implementaitons. 

If special treatment of Span in foreach goes into spec, then it becomes a language feature that every confirming compiler needs to provide. The spec will need to have detailed answers on all the questions about what makes the type an "foreach-able span type" and how the iteration type is inferred  (it is the type parameter or return type of the indexer? what if indexer does not exist? what if GetEnumerator exists and conflicts, who wins? etc...) and it will need to provide directions on how such special case of foreach should execute.

Current spec for `foreach` is 3 pages of terse legalize and yet does not mention special types like `string`. That is because the pattern is sufficient to make `string` foreach-able and answer all the needed questions. (compiler may emit more efficient code where possible and is free to tweak that as needed)

Duplicating `foreach` spec for Span and ReadOnlySpan is essentially  "bloating C# spec" and perpetuating coupling of language with a lot of details of Span and ReadOnlySpan API, when we already have a sufficient solution.

I am sure there are other ways to ensure that when `GetEnumerator` is not called by the user code it does not end up in AOT images.

**Comment by KrzysztofCwalina** [comment]
> I am sure there are other ways to ensure that when GetEnumerator is not called by the user code it does not end up in AOT images.

@jkotas, are there practical ways to do it?

And, @VSadov, I was not proposing that we special case Span/ROSpan. I though we would augment the spec with a general pattern based on Length/this[int] that does not require GetEnumerator.

**Comment by jkotas** [comment]
Cheap? Not really - given how the codegen and runtime is designed today. Possible with enough effort? Yes.

As I have said in https://github.com/dotnet/csharplang/issues/1085#issuecomment-342663170, the Span/Memory are likely to be used with relatively small number of instantiations, so I am not terribly concerned about the enumerator code bloat. The damage from ValueTuple is likely going to be much higher.

The pattern of C# generating complex IL for simple source code, and then asking the JIT to cleanup the mess and turn in into something fast, small and simple again at runtime, is less than ideal. It is what leads to idea of offloading the job of cleaning up the mess into a separate IL optimizer tool.

**Comment by VSadov** [comment]
We will generate good code for this case. I just so not believe there is enough reason to add to the language spec in this case. There are short and long term costs to that too.

A also think span iterators could be farly useful abstractions on their own.

**Comment by KrzysztofCwalina** [comment]
Span iterator is a by ref struct, i.e. not a great abstraction :-)

**Comment by VSadov** [comment]
Yeah, probably not an _abstraction_ :-) 
There could be still some use beyond `foreach`. Like `Parse(ref iterator)` may look a bit cleaner than `Parse(span, ref currentPosition)`.
Or if you need to iterate two spans at the same time...  Any time you need to package a span and a cursor. Not sure how common though.

I hope noone writes something Linq-like for these guys ;-)

**Comment by Pzixel** [comment]
@VSadov 
> I hope noone writes something Linq-like for these guys ;-)

This is exactly what I were thinking about, because why not? We are often taking some array/list as parameter and then process it via linq so we cannot change `T[]` with `Span<T>` if they do not support it. 

It's probably an abuse instead of fair use, but I don't see why it's unwanted.

**Comment by fanoI** [comment]
Guys but why is not possible to "transform" foreach in for always or at least in more cases? It is done for arrays and strings but for example not for Lists?

Why when I do:

```cs
foreach(var x in MyList)
```

I don't get automatically:
```cs
for(var i; i < MyList.count; i++)
```
I mean why it is not the "first" thing done? Why create all the enumerator the try / catch / finally and so on... when there a simple for solution...

I want not read on forums anymore that between the optimization tricks there is "never use foreach" :-)
From the user point of view it seems really a nice "for" he cannot imagine what is behind it...

**Comment by CyrusNajmabadi** [comment]
> I mean why it is not the "first" thing done? Why create all the enumerator the try / catch / finally and so on... when there a simple for solution...

Because that may have different semantics from the original code.  For example, if there is an IDisposable method to call, it won't be called. 

Note that for List<T> this code path is already optimized as a struct-enumerator is created.  So there is no actual cost of allocating that enumerator, and there are no virtual calls involved.

**Comment by fanoI** [comment]
While there is a struct Enumerator for List, replacing foreach with for you get:

> The application ran for 0.924s after the change, leading to a 65% reduction in time over the original by avoiding the enumerator.

In this special case as we have no IDisposable (we are using a struct as Enumerator) what will be this "semantic" difference from the user point of view? IMHO the only observable thing will be that is 65% faster that will be always good, right?

I imagine that it is not "pretty" to do this in the compiler but you had done already this for array, String and you must do same thing for List (in a lot of cases used as a "resizable" array) and obviously for Span that is tough for fast / high performance code and have to create an Enumerable only for do "foreach" will defeat its purpose.

**Comment by CyrusNajmabadi** [comment]
> In this special case as we have no IDisposable (we are using a struct as Enumerator) 

There is a Dispose method:

```c#
public class List<T> : ...
{
...
        public struct Enumerator : IEnumerator<T>, IDisposable, IEnumerator
        {
            public T Current { get; }

            public void Dispose();
            public bool MoveNext();
        }
}
```

**Comment by CyrusNajmabadi** [comment]
> I imagine that it is not "pretty" to do this in the compiler but you had done already this for array, String

Array and string are defined by the runtime.  List<T> is not.  We may be running with an implementation for which there are different semantics here.

**Comment by CyrusNajmabadi** [comment]
>  that is tough for fast / high performance code and have to create an Enumerable only for do "foreach" will defeat its purpose.

That's not true at all.  One of hte major things spans are trying to address is overhead of heap allocations, and the difficulty of doing things fast while avoiding unsafe code.  

'foreach' doesn't make things slow, though it may make things slightly slower.  The code you linked to is a microbenchmark that is pretty uninteresting.  Even working with lists of 100s of millions of entries, i measure barely any difference between the two.  Seriously, it's around 22ms.  :) 

The goal is not to make microbenchmarks fast.  It's to make real world code fast and expressive.  Efforts like Span do that.  Now, if i was in a realm where those milliseconds mattered, i still wouldn't want this compiler feature.  Why?  Because, at that point, i wouldn't even be using List<T> anymore.  I wouldn't even be using an array.  I'd just drop down to raw pointers (which are around 3ms faster in my measurement).

This is the issue with chasing raw microbenchmark numbers.  They simply lead you to a space where you can't use anyting in the language, as there will always be some uglier way to do it faster.

**Comment by fanoI** [comment]
Well if I'm looking at the correct code The IEnumerable of List has Dispose() only to satisty the interface is empty indeed as no memory has been allocated: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Collections/Generic/List.cs#L1188

However of the blog posts is wrong (was 2.667s with foreach vs 0.924s that is 1.703s of difference) or well foreach on List in not the same thing of a for... in any case the generated code is more complex, it expands to a try / finally in which is called a Dispose() that well in the end does nothing!

On this I surely concord:
> This is the issue with chasing raw microbenchmark numbers. They simply lead you to a space where you can't use anyting in the language, as there will always be some uglier way to do it faster.

better to write high level C# code but here we have a 1.703s of difference respect to the "ugly" version if the test has been done correctly!

**Comment by Pzixel** [comment]
@fanoI you are testing current list implementation on current CLR. Things may change and it doesn't worth to implement some "optimizations" that could be the opposite later.

**Comment by CyrusNajmabadi** [comment]
>  in any case the generated code is more complex, it expands to a try / finally in which is called a Dispose() that well in the end does nothing!

There is no way for the C# compiler to know that the Dispose does nothing.  You may have an implementation of Dispose that does important work.

**Comment by CyrusNajmabadi** [comment]
> it expands to a try / finally in which is called a Dispose() that well in the end does nothing!

Note: it would be interesting if the runtime detected that a finally was a no-op.  i.e. because it calls a method which, when inlined, does nothing.  I wonder if the try/finally could be elided entirely at that point.  

Possibly something you could investigate given your interest here?

**Comment by AndyAyersMS** [comment]
FWIW the jit can see that the dispose does nothing, and further that a try with empty finally can be optimized away, so that bit of abstraction "overhead" is typically not there in the generated code.

**Comment by CyrusNajmabadi** [comment]
> so that bit of abstraction "overhead" is typically not there in the generated code.

Nice!

Out of curiosity andy, do we know why the struct-enumerator form of foreach over a list is poorer performing than the indexing form?  Is it simply the 'version' check?  Is the size of the method too large to inline?  Thanks!

**Comment by CyrusNajmabadi** [comment]
> I imagine that it is not "pretty" to do this in the compiler but you had done already this for array, String and you should do same thing for List

Note: replacing the existing GetEnumerator form with straight indexing *will* be a change in behavior.  With the existing GetEnumerator logic, you get the nice behavior that mutations to the underlying collection lead to exceptions happening while iterating.  That would not happen with the indexing form.  This is a change in behavior, and having the compiler make this optimization would break the code that the BCL intentionally put into their enumeration logic.

**Comment by AndyAyersMS** [comment]
I don't know offhand, let me take a look.

**Comment by AndyAyersMS** [comment]
Yes, the version check makes `Enumerator.MoveNext` complicated enough that the jit won't inline it. We have these inner loops (here simply summing the contents of a `List<int>`):
```asm
;; for loop
Inlines into 06000001 P:SumFor(ref):int
  [1 IL=0009 TR=000022 06003ABF] [profitable inline] System.Collections.Generic.List`1[Int32][System.Int32]:get_Item(int):int:this
    [0 IL=0009 TR=000051 060001CA] [FAILED: does not return] System.ThrowHelper:ThrowArgumentOutOfRange_IndexException()
  [2 IL=0022 TR=000012 06003AB9] [below ALWAYS_INLINE size] System.Collections.Generic.List`1[Int32][System.Int32]:get_Count():int:this

G_M56821_IG03:
       413BD0               cmp      edx, r8d
       731E                 jae      SHORT G_M56821_IG06

G_M56821_IG04:
       4C8B4908             mov      r9, gword ptr [rcx+8]
       413B5108             cmp      edx, dword ptr [r9+8]
       731A                 jae      SHORT G_M56821_IG07
       4C63D2               movsxd   r10, edx
       4303449110           add      eax, dword ptr [r9+4*r10+16]
       FFC2                 inc      edx
       413BD0               cmp      edx, r8d
       7CE2                 jl       SHORT G_M56821_IG03

;; foreach
Inlines into 06000002 P:SumForeach(ref):int
  [1 IL=0003 TR=000006 06003AE0] [below ALWAYS_INLINE size] System.Collections.Generic.List`1[Int32][System.Int32]:GetEnumerator():struct:this
    [2 IL=0001 TR=000056 06004A2E] [profitable inline] Enumerator[Int32][System.Int32]:.ctor(ref):this
  [3 IL=0013 TR=000025 06004A32] [below ALWAYS_INLINE size] Enumerator[Int32][System.Int32]:get_Current():int:this
  [0 IL=0025 TR=000017 06004A30] [FAILED: noinline per IL/cached result] Enumerator[Int32][System.Int32]:MoveNext():bool:this
  [4 IL=0042 TR=000045 06004A2F] [below ALWAYS_INLINE size] Enumerator[Int32][System.Int32]:Dispose():this

G_M54913_IG04:
       8B4C2430             mov      ecx, dword ptr [rsp+30H]
       03F1                 add      esi, ecx
       488D4C2420           lea      rcx, bword ptr [rsp+20H]
       E8A021B75E           call     Enumerator[Int32][System.Int32]:MoveNext():bool:this
       85C0                 test     eax, eax
       75EC                 jne      SHORT G_M54913_IG04
```
If we managed to inline both `MoveNext` and its callee `MoveNextRare` I would expect perf of `foreach` would come a good deal closer perf of `for`, with still some cost for the extra safety checks (though in a simple case like this the jit might also be able to optimize those out).

Note the `for` case is not very well optimized either. The jit has no way of knowing the relationship of `List.Count` and the array size, so we have an array bounds check per iteration, and for some reason a redundant loop exit check.

As noted above  `Dispose` is inlined and try/finally is removed, so the finally overhead in the `foreach` case is gone.

**Comment by CyrusNajmabadi** [comment]
> Yes, the version check makes Enumerator.MoveNext complicated enough that the jit won't inline it. We have these inner loops (here simply summing the contents of a List<int>):

It would be interesting to see if special casing the inlining specifically for the List enumeration case had any actual benefits on the corpus of perf tests the runtime has.

I know that JS engines will often time have extremely special cased logic for certain patterns or APIs. While it's nice to have a generalized system, some specific APIs are used in such a common way, that optimizations may help.  Of course, the JS sphere is totally different as well.  There, the difference between having and not having these optimizations can be orders of magnitude perf changes.  Here, it seems like it might have a little boost.  But it's unclear if that would lead to real world benefits in practical scenarios.

**Comment by jkotas** [comment]
> The jit has no way of knowing the relationship of List.Count and the array size, so we have an array bounds check per iteration.

There is no guaranteed relationship. It can be arbitrary if the user code has a race condition. We generally build the runtime and framework to guarantee no buffer overruns, even if the user code has race condition.

**Comment by AndyAyersMS** [comment]
It is easy enough to add the aggressive inline attributes here. Doing this gives the following innerloop code for ` foreach`:
```asm
Inlines into 06000002 P:SumForeach(ref):int
  [1 IL=0003 TR=000006 06003AE0] [below ALWAYS_INLINE size] System.Collections.Generic.List`1[Int32][System.Int32]:GetEnumerator():struct:this
    [2 IL=0001 TR=000058 06004A2E] [profitable inline] Enumerator[Int32][System.Int32]:.ctor(ref):this
  [3 IL=0013 TR=000027 06004A32] [below ALWAYS_INLINE size] Enumerator[Int32][System.Int32]:get_Current():int:this
  [4 IL=0025 TR=000017 06004A30] [aggressive inline attribute] Enumerator[Int32][System.Int32]:MoveNext():bool:this
    [5 IL=0075 TR=000120 06004A31] [aggressive inline attribute] Enumerator[Int32][System.Int32]:MoveNextRare():bool:this
      [0 IL=0019 TR=000204 060001ED] [FAILED: does not return] System.ThrowHelper:ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion()

G_M54916_IG04:
       4103C2               add      eax, r10d

G_M54916_IG05:
       443BC1               cmp      r8d, ecx
       7523                 jne      SHORT G_M54916_IG06
       443B4A18             cmp      r9d, dword ptr [rdx+24]
       731D                 jae      SHORT G_M54916_IG06
       4C8B5208             mov      r10, gword ptr [rdx+8]
       453B4A08             cmp      r9d, dword ptr [r10+8]
       7335                 jae      SHORT G_M54916_IG11
       4D63D9               movsxd   r11, r9d
       478B549A10           mov      r10d, dword ptr [r10+4*r11+16]
       41FFC1               inc      r9d
       41BB01000000         mov      r11d, 1
       EB12                 jmp      SHORT G_M54916_IG08

G_M54916_IG06:
       443BC1               cmp      r8d, ecx
       7517                 jne      SHORT G_M54916_IG10

G_M54916_IG07:
       448B4A18             mov      r9d, dword ptr [rdx+24]
       41FFC1               inc      r9d
       4533D2               xor      r10d, r10d
       4533DB               xor      r11d, r11d

G_M54916_IG08:
       4585DB               test     r11d, r11d
       75BE                 jne      SHORT G_M54916_IG04
```
The jit did not clean up the version check artifacts so there are extra tests for those, plus the extra code for the `MoveNextRare` case. Plus the range check (which is also there in the `for` version). Roughly speaking this makes `foreach` about 1.3x slower than `for` (vs ~ 2.9x w/o the inlining).

Not clear why the jit can't remove the extra checks. I'll open an issue to investigate. But without doing so we see about 40 bytes of extra code from the forced inlines (and even if we enhance the jit, these won't always be able to be optimized away, if the per-element computation does something opaque like call a method).

**Comment by CyrusNajmabadi** [comment]
> It is easy enough to add the aggressive inline attributes here.

Sounds like a potentially worthwhile change for a coding pattern that we would want to be as streamlined as possible. :)  Would you want to drive making that change to the BCL?

**Comment by AndyAyersMS** [comment]
Let me look into some of the jit issues first. If we can get at this without some of the code bloat then it is easier to justify.

**Comment by VSadov** [comment]
BTW. The version check artifacts is one of the reasons C# does not optimize `foreach` into indexing with `List<T>`.  
The version check is internal to the List, so it is not possible to do such translation at compiler level without observably changing the behavior.

**Comment by fanoI** [comment]
This discussion is really interesting for me! So it is this the semantic difference: if while I'm doing a for in a List another thread changes it adding / removing an element the for() can "see" the change while the foreach() you will get an exception!

It is this cure of little details that I like of .NET if you optimize foreach() in a for() you get a different behavior and slighter unsafe code!

In any case I'm happy that my question of the issue has made you see that were space for optimizations https://github.com/dotnet/coreclr/issues/15130, it is for this the vtune amplifier is a really interesting tool (but the solution of replace foreach() with for() is wrong at this point and not a really solution),

**Comment by jcouv** [comment]
@stephentoub From discussion with @VSadov, this doesn't require a language decision. He has a PR to implement the optimization without affecting the language spec.
Close the issue if ok with that.

**Comment by stephentoub** [comment]
> He has a PR to implement the optimization without affecting the language spec.

I see the PR at https://github.com/dotnet/roslyn/pull/23363.  With that, can we remove the Span.Enumerator type, or do we need to keep that even though it'll be unused?

**Comment by VSadov** [comment]
`Span.Enumerator` is a part of `foreach` pattern. That is how the type claims that it can do `foreach` . 

Also a compiler that does not perform optimization as in the dotnet/roslyn#23363, will actually use the enumerator. So it would be better if `Span.Enumerator` has reasonable implementation.

**Comment by AceHack** [comment]
Is it a way to use linq like any, all, select, where on span?

**Comment by CyrusNajmabadi** [comment]
> Is it a way to use linq like any, all, select, where on span?

This would be a question for the runtime, not the language.  the lnaguage has no problem using those linq methods on span *provided* that it can find appropriate extension methods it can call into.  So, if you (or some other library) provides those extensions, this will "just work (tm)".  But, as before, it's not part of the language for that to work.  Thanks!
